File: BOJ/BOJ/Gold4/boj 30893 트리 게임.py
import sys 

N, S, E = map(int,input().split())

# S를 트리의 루트로 가정.
# dfs로 경로 탐색.
# 1턴부터 시작
# 짝수 턴에 갈 수 있는 길이 하나 뿐이면 선공, 아니면 후공

tree = [[] for _ in range(N+1)]

for _ in range(N-1):
    a,b = map(int,input().split())
    tree[a].append(b)
    tree[b].append(a)

st = [S]
prev = [None for _ in range(N+1)]
prev[S] = -1
path_cnt = [0 for _ in range(N+1)]
while st:
    current = st.pop()
    if current == E:
        break
    for nxt in tree[current]:
        if prev[nxt] is not None:
            continue
        st.append(nxt)
        prev[nxt] = current
        path_cnt[current] += 1
        # print(f"{current} -> {nxt}")

def get_path():
    res = []
    c = E
    while c != S:
        res.append(c)
        c = prev[c]
    return res

path_without_S = list(reversed(get_path()))

# for i,e in enumerate(path_without_S):
#     print(f"{i} : {e}")
#     print(f"cnt : {path_cnt[e]}")
for n in path_without_S[:-1:2]:
    # print(f"checked {n}")
    if path_cnt[n] > 1:
        print("Second")
        break
else:
    print("First")

File: BOJ/BOJ/Gold4/boj 30893 트리게임.cpp
#include <iostream>
#include <vector>

using namespace std;

void fastio() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
}


int N, S, E;

vector<int> tree[100'001];
int prev_node[100'001] = { 0, };

vector<int> findPath() {
    vector<int> result;
    int current = E;
    while (current != S) {
        result.push_back(current);
        current = prev_node[current];
    }
    return result;
}

bool dfs(int node) {
    if (node == E) {
        return true;
    }
    for (int nxt : tree[node]) {
        if (prev_node[nxt] != 0)
            continue;
        prev_node[nxt] = node;
        if (dfs(nxt))
            return true;
    }
    return false;
}

int main() {
    fastio();
    cin >> N >> S >> E;
    for (size_t i = 0; i < N - 1; i++)
    {
        int a, b;
        cin >> a >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    prev_node[S] = -1;
    dfs(S);
    vector<int> reversed_path = findPath();
    for (int i = reversed_path.size() - 1; i > 0; i -= 2) {
        int n = reversed_path[i];
        if (tree[n].size() - 1 > 1) {
            cout << "Second" << "\n";
            return 0;
        }
    }
    cout << "First" << "\n";

    return 0;
}